File: WebMock/src/core/generate.ts
// 領域拡張（Region Growing）による分割と、プレイ開始時の初期散らし（RULES.md 3.2）。
// Three.js に依存しない純粋なロジック。乱数は必ず createRng を通す（Math.random を使わない）。

import {
  addVec3,
  compareVec3,
  IDENTITY_ORIENTATION,
  ORIENTATION_COUNT,
  vec3,
  vec3Key,
  type Vec3,
} from './grid';
import { createPiece, localOrigin, placedVoxels, type Piece, type Placement } from './piece';
import { createRng, type Rng } from './rng';
import { isSolved } from './solve';

/** 空間サイズ N の下限（RULES.md 3.1）。 */
export const MIN_SPACE_SIZE = 3;
/** 空間サイズ N の上限（RULES.md 3.1）。 */
export const MAX_SPACE_SIZE = 7;
/** 分割数 M の下限（RULES.md 3.1）。 */
export const MIN_PIECE_COUNT = 2;

// 未割り当てのボクセルを表す owner の値。
const UNASSIGNED = -1;

// 散らしで 1 ピースを置くときのリトライ上限と、範囲を広げる間隔。
const SCATTER_ATTEMPT_LIMIT = 10000;
const SCATTER_WIDEN_INTERVAL = 100;
// 散らし全体（クリア状態を避けるための引き直し）の上限。
const SCATTER_RETRY_LIMIT = 8;

/** 6 近傍のオフセット。 */
const NEIGHBOR_OFFSETS: readonly Vec3[] = [
  vec3(1, 0, 0),
  vec3(-1, 0, 0),
  vec3(0, 1, 0),
  vec3(0, -1, 0),
  vec3(0, 0, 1),
  vec3(0, 0, -1),
];

/**
 * 生成結果。
 * pieces は局所座標へ正規化済み（局所原点が (0,0,0)）で、solution は各ピースの解答配置。
 * pieces[i] と solution[i] は同じピース（id は i）を指す。
 */
export type GeneratedPuzzle = {
  readonly n: number;
  readonly m: number;
  readonly seed: number;
  readonly pieces: readonly Piece[];
  readonly solution: readonly Placement[];
};

/** 配列の要素を取り出す。範囲外は実装のバグなので例外にする（noUncheckedIndexedAccess 対策）。 */
function at<T>(array: readonly T[], index: number): T {
  const value = array[index];
  if (value === undefined) throw new Error(`配列の範囲外アクセス: ${index}`);
  return value;
}

/** N が 3..7 の整数であることを確かめる。 */
function assertSpaceSize(n: number): void {
  if (!Number.isInteger(n) || n < MIN_SPACE_SIZE || n > MAX_SPACE_SIZE) {
    throw new RangeError(
      `空間サイズ N は ${MIN_SPACE_SIZE}..${MAX_SPACE_SIZE} の整数 (got ${String(n)})`,
    );
  }
}

/** シードが整数であることを確かめる。 */
function assertSeed(seed: number): void {
  if (!Number.isInteger(seed)) throw new RangeError(`シードは整数 (got ${String(seed)})`);
}

/**
 * 空間サイズ N に対して選べる分割数 M の上限（RULES.md 3.1）。
 *
 * M のプリセットは N から N−2 刻みの 5 段なので、上限はその最大値 N + 4(N−2) = 5N − 8
 * （N=3 → 7、N=4 → 12、N=5 → 17、N=6 → 22、N=7 → 27）。
 */
export function maxPieces(n: number): number {
  assertSpaceSize(n);
  return 5 * n - 8;
}

/** M が 2..maxPieces(n) の整数であることを確かめる。 */
function assertPieceCount(n: number, m: number): void {
  const limit = maxPieces(n);
  if (!Number.isInteger(m) || m < MIN_PIECE_COUNT || m > limit) {
    throw new RangeError(`分割数 M は ${MIN_PIECE_COUNT}..${limit} の整数 (N=${n}, got ${String(m)})`);
  }
}

/** グリッド座標 → 一次元インデックス。 */
function encodeIndex(v: Vec3, n: number): number {
  return (v.x * n + v.y) * n + v.z;
}

/** 一次元インデックス → グリッド座標。 */
function decodeIndex(index: number, n: number): Vec3 {
  return vec3(Math.floor(index / (n * n)), Math.floor(index / n) % n, index % n);
}

/** 座標が N×N×N の中にあるか。 */
function isInside(v: Vec3, n: number): boolean {
  return v.x >= 0 && v.x < n && v.y >= 0 && v.y < n && v.z >= 0 && v.z < n;
}

/**
 * 領域拡張で N×N×N を M 個の連結なピースに分割する（RULES.md 3.2）。
 * 返す pieces は局所座標に正規化済みで、絶対座標（解答位置）は solution が持つ。
 */
export function generatePuzzle(n: number, m: number, seed: number): GeneratedPuzzle {
  assertSpaceSize(n);
  assertPieceCount(n, m);
  assertSeed(seed);

  const rng = createRng(seed);
  const total = n * n * n;
  // owner[index] = そのボクセルを持つピース id（未割り当ては UNASSIGNED）
  const owner: number[] = new Array<number>(total).fill(UNASSIGNED);
  // members[piece] = そのピースのボクセルのインデックス
  const members: number[][] = Array.from({ length: m }, (): number[] => []);
  // frontiers[piece] = 吸収候補（隣接する未割り当てボクセル）。割り当て済みの古い候補も混ざる
  const frontiers: number[][] = Array.from({ length: m }, (): number[] => []);

  /** ボクセルをピースに吸収し、その 6 近傍の空きを候補に積む。 */
  const absorb = (index: number, piece: number): void => {
    owner[index] = piece;
    at(members, piece).push(index);
    const v = decodeIndex(index, n);
    const frontier = at(frontiers, piece);
    for (const offset of NEIGHBOR_OFFSETS) {
      const neighbor = addVec3(v, offset);
      if (!isInside(neighbor, n)) continue;
      const neighborIndex = encodeIndex(neighbor, n);
      if (at(owner, neighborIndex) !== UNASSIGNED) continue;
      frontier.push(neighborIndex);
    }
  };

  // 2. 相異なるランダム座標に M 個のシードを置く。
  //    部分 Fisher-Yates で先頭 M 個を選ぶので、重複のリトライ無しに必ず相異なる。
  const shuffled: number[] = Array.from({ length: total }, (_, i): number => i);
  for (let i = 0; i < m; i++) {
    const j = i + rng.nextInt(total - i);
    const a = at(shuffled, i);
    const b = at(shuffled, j);
    shuffled[i] = b;
    shuffled[j] = a;
    absorb(b, i);
  }

  // 3. 各ピースが隣接する空きを 1 つ吸収する、をラウンドロビンで繰り返す。
  let assigned = m;
  while (assigned < total) {
    let grew = false;
    for (let piece = 0; piece < m && assigned < total; piece++) {
      const frontier = at(frontiers, piece);
      let picked = UNASSIGNED;
      while (frontier.length > 0) {
        const k = rng.nextInt(frontier.length);
        const candidate = at(frontier, k);
        frontier[k] = at(frontier, frontier.length - 1);
        frontier.pop();
        if (at(owner, candidate) === UNASSIGNED) {
          picked = candidate;
          break;
        }
      }
      // 空きが無いピースはこのラウンドをスキップ
      if (picked === UNASSIGNED) continue;
      absorb(picked, piece);
      assigned++;
      grew = true;
    }
    if (grew) continue;

    // すべてのピースがスキップになったのに空きが残る場合（RULES.md 3.2-3 の孤立した空き）。
    // 候補は吸収されるまで frontier に残る作りなので実際にはここへ来ないが、仕様どおり保険を置く。
    let remaining: number[] = [];
    for (let index = 0; index < total; index++) {
      if (at(owner, index) === UNASSIGNED) remaining.push(index);
    }
    while (remaining.length > 0) {
      const next: number[] = [];
      let progressed = false;
      for (const index of remaining) {
        const v = decodeIndex(index, n);
        let host = UNASSIGNED;
        for (const offset of NEIGHBOR_OFFSETS) {
          const neighbor = addVec3(v, offset);
          if (!isInside(neighbor, n)) continue;
          const neighborOwner = at(owner, encodeIndex(neighbor, n));
          if (neighborOwner !== UNASSIGNED) {
            host = neighborOwner;
            break;
          }
        }
        // どのピースにも隣接しない空きは次の周回に回す（周りが埋まれば吸収できる）
        if (host === UNASSIGNED) {
          next.push(index);
          continue;
        }
        absorb(index, host);
        assigned++;
        progressed = true;
      }
      if (!progressed) throw new Error('生成に失敗: どのピースにも隣接しない空きが残った');
      remaining = next;
    }
  }

  // 4. 解答位置（絶対座標）から局所座標へ正規化し、正規化で引いたオフセットを解答配置にする。
  const pieces: Piece[] = [];
  const solution: Placement[] = [];
  for (let id = 0; id < m; id++) {
    const absolute = at(members, id)
      .map((index): Vec3 => decodeIndex(index, n))
      .sort(compareVec3);
    pieces.push(createPiece(id, absolute));
    solution.push({
      pieceId: id,
      orientation: IDENTITY_ORIENTATION,
      position: localOrigin(absolute),
    });
  }

  return { n, m, seed, pieces, solution };
}

/** 領域拡張で分割したピース（局所座標に正規化済み）。解答位置は solutionPlacements で得る。 */
export function generatePieces(n: number, m: number, seed: number): Piece[] {
  return generatePuzzle(n, m, seed).pieces.slice();
}

/**
 * 各ピースの解答配置（向きは恒等、位置は生成時の絶対座標のオフセット）。
 *
 * 解釈: 正規化でオフセットが落ちるため Piece だけからは解答位置を復元できない。
 * そこで generatePuzzle が pieces と solution を対で返す形にし、この関数は同じ引数で
 * 解答配置だけを取り出す薄い入口とする（両方要るときは generatePuzzle を 1 回呼べばよい）。
 */
export function solutionPlacements(n: number, m: number, seed: number): Placement[] {
  return generatePuzzle(n, m, seed).solution.slice();
}

/** 散らしのオプション。既定は「向きはランダム・固定ピース無し」で、従来の 3 引数の呼び出しと同じ。 */
export type ScatterOptions = {
  /** 向きをランダムにするか。false なら全ピースを IDENTITY_ORIENTATION で置く。既定 true */
  readonly allowRotation?: boolean;
  /** 散らさずにそのまま残す配置（ヒントで固定したピース）。既定は空 */
  readonly keep?: readonly Placement[];
};

/** keep をピース id 引きの表にする。未知の id / 重複した id は呼び出し側のバグなので例外。 */
function buildKeepMap(
  pieces: readonly Piece[],
  keep: readonly Placement[] | undefined,
): ReadonlyMap<number, Placement> {
  const map = new Map<number, Placement>();
  if (keep === undefined) return map;
  const known = new Set(pieces.map((piece): number => piece.id));
  for (const placement of keep) {
    if (!known.has(placement.pieceId)) {
      throw new Error(`散らし: keep に未知のピース id ${placement.pieceId}`);
    }
    if (map.has(placement.pieceId)) {
      throw new Error(`散らし: keep のピース id ${placement.pieceId} が重複している`);
    }
    map.set(placement.pieceId, placement);
  }
  return map;
}

/**
 * 1 回分の散らし。重なりが出たらその場でリトライし、詰まったら範囲を広げる。
 * keepById にある配置はそのまま採用し、そのボクセルは先に占有として登録する。
 */
function scatterOnce(
  pieces: readonly Piece[],
  n: number,
  rng: Rng,
  allowRotation: boolean,
  keepById: ReadonlyMap<number, Placement>,
): Placement[] {
  const occupied = new Set<string>();
  const placements: Placement[] = [];
  // 立方体は [0, N-1]³ にあるので、その中心のまわり ±(N+2) を既定の散らし範囲にする
  const center = Math.floor((n - 1) / 2);

  // 固定ピースのボクセルを先に占有として登録し、残りがそこへ重ならないようにする
  for (const piece of pieces) {
    const kept = keepById.get(piece.id);
    if (kept === undefined) continue;
    for (const v of placedVoxels(piece, kept)) occupied.add(vec3Key(v));
  }

  for (const piece of pieces) {
    const kept = keepById.get(piece.id);
    if (kept !== undefined) {
      // そのまま残す。乱数も消費しないので keep の内容が同じなら結果も同じ
      placements.push(kept);
      continue;
    }

    let half = n + 2;
    let attempts = 0;
    for (;;) {
      const span = half * 2 + 1;
      const placement: Placement = {
        pieceId: piece.id,
        orientation: allowRotation ? rng.nextInt(ORIENTATION_COUNT) : IDENTITY_ORIENTATION,
        position: vec3(
          center - half + rng.nextInt(span),
          center - half + rng.nextInt(span),
          center - half + rng.nextInt(span),
        ),
      };
      const keys = placedVoxels(piece, placement).map(vec3Key);
      if (keys.every((key): boolean => !occupied.has(key))) {
        for (const key of keys) occupied.add(key);
        placements.push(placement);
        break;
      }
      attempts++;
      if (attempts >= SCATTER_ATTEMPT_LIMIT) {
        throw new Error(`散らしに失敗: ピース ${piece.id} を置けなかった`);
      }
      // 詰まってきたら範囲を広げ、必ず終わるようにする
      if (attempts % SCATTER_WIDEN_INTERVAL === 0) half++;
    }
  }
  return placements;
}

/**
 * プレイ開始時の初期散らし（RULES.md 3.2-5）。
 * 各ピースにランダムな向き（24 通り）と、立方体の周囲 ±(N+2) 程度のランダムな位置を与える。
 * ピース同士は重ならない。同じ引数なら常に同じ配置になる。
 *
 * options で「向きを恒等に固定する（難易度: 回転なし）」「指定した配置は散らさず残す（ヒントで
 * 固定したピース）」を指定できる。返り値は常に全ピース分で、並びは pieces の並びに揃う。
 *
 * 解釈: 散らした直後にクリア判定が真になると開始と同時にクリアしてしまうので、真なら引き直す。
 * ただし keep があるときは固定の進み具合によっては避けようが無いので、上限まで引き直しても
 * 避けられなければ例外にせず最後の配置を返す。
 */
export function scatterPlacements(
  pieces: readonly Piece[],
  n: number,
  seed: number,
  options?: ScatterOptions,
): Placement[] {
  assertSpaceSize(n);
  assertSeed(seed);
  if (pieces.length === 0) throw new RangeError('散らし: ピースが空');

  const allowRotation = options?.allowRotation ?? true;
  const keepById = buildKeepMap(pieces, options?.keep);

  const rng = createRng(seed);
  let last: Placement[] | null = null;
  for (let retry = 0; retry < SCATTER_RETRY_LIMIT; retry++) {
    const placements = scatterOnce(pieces, n, rng, allowRotation, keepById);
    if (!isSolved(pieces, placements, n)) return placements;
    last = placements;
  }
  if (keepById.size > 0 && last !== null) return last;
  throw new Error('散らしに失敗: クリア状態でない配置を作れなかった');
}

File: Source/CUBELITHCore/Private/Generate.cpp
#include "Generate.h"

#include "Algo/StableSort.h"
#include "Rng.h"
#include "Solve.h"

// 解釈: TS の at()（範囲外アクセスを例外にするヘルパ）は noUncheckedIndexedAccess 対策なので移植しない。
// C++ では TArray の operator[] がデバッグビルドで添字を検査する。

namespace Cubelith
{
	namespace
	{
		/** 未割り当てのボクセルを表す Owner の値（TS の UNASSIGNED） */
		constexpr int32 Unassigned = -1;

		/**
		 * 6 近傍のオフセット（TS の NEIGHBOR_OFFSETS）。
		 * 候補を積む順が以降の乱数の当たり先を決めるので、並び（+X, -X, +Y, -Y, +Z, -Z）を変えない。
		 */
		constexpr FVec3 NeighborOffsets[] = {
			{ 1, 0, 0 }, { -1, 0, 0 }, { 0, 1, 0 }, { 0, -1, 0 }, { 0, 0, 1 }, { 0, 0, -1 } };

		/** N が 3..7 であることを確かめる（TS の assertSpaceSize）。整数かどうかは C++ では型で決まる */
		void AssertSpaceSize(int32 N)
		{
			checkf(N >= MinSpaceSize && N <= MaxSpaceSize, TEXT("空間サイズ N は %d..%d の整数 (got %d)"),
				MinSpaceSize, MaxSpaceSize, N);
		}

		/** M が 2..MaxPieces(N) であることを確かめる（TS の assertPieceCount） */
		void AssertPieceCount(int32 N, int32 M)
		{
			const int32 Limit = MaxPieces(N);
			checkf(M >= MinPieceCount && M <= Limit, TEXT("分割数 M は %d..%d の整数 (N=%d, got %d)"),
				MinPieceCount, Limit, N, M);
		}

		/** グリッド座標 → 一次元インデックス（TS の encodeIndex） */
		int32 EncodeIndex(const FVec3& V, int32 N)
		{
			return (V.X * N + V.Y) * N + V.Z;
		}

		/**
		 * 一次元インデックス → グリッド座標（TS の decodeIndex）。
		 * C++ の整数の除算は 0 方向への切り捨てで TS の Math.floor と違うが、Index は常に非負なのでそのままでよい。
		 */
		FVec3 DecodeIndex(int32 Index, int32 N)
		{
			return Vec3(Index / (N * N), (Index / N) % N, Index % N);
		}

		/** 座標が N×N×N の中にあるか（TS の isInside） */
		bool IsInside(const FVec3& V, int32 N)
		{
			return V.X >= 0 && V.X < N && V.Y >= 0 && V.Y < N && V.Z >= 0 && V.Z < N;
		}

		/** 散らしで 1 ピースを置くときのリトライ上限（TS の SCATTER_ATTEMPT_LIMIT） */
		constexpr int32 ScatterAttemptLimit = 10000;

		/** 散らしで範囲を広げる間隔（TS の SCATTER_WIDEN_INTERVAL） */
		constexpr int32 ScatterWidenInterval = 100;

		/** 散らし全体（クリア状態を避けるための引き直し）の上限（TS の SCATTER_RETRY_LIMIT） */
		constexpr int32 ScatterRetryLimit = 8;

		/**
		 * Keep をピース id 引きの表にする（TS の buildKeepMap）。未知の id / 重複した id は呼び出し側のバグなので checkf。
		 * 引くだけの用途なので TMap でよい（この表を回って順序に依存することはしない）。
		 */
		TMap<int32, FPlacement> BuildKeepMap(TArrayView<const FPiece> Pieces, TArrayView<const FPlacement> Keep)
		{
			TMap<int32, FPlacement> Map;
			if (Keep.Num() == 0)
			{
				return Map;
			}

			TSet<int32> Known;
			Known.Reserve(Pieces.Num());
			for (const FPiece& Piece : Pieces)
			{
				Known.Add(Piece.Id);
			}

			Map.Reserve(Keep.Num());
			for (const FPlacement& Placement : Keep)
			{
				checkf(Known.Contains(Placement.PieceId), TEXT("散らし: keep に未知のピース id %d"), Placement.PieceId);
				checkf(!Map.Contains(Placement.PieceId), TEXT("散らし: keep のピース id %d が重複している"), Placement.PieceId);
				Map.Add(Placement.PieceId, Placement);
			}
			return Map;
		}

		/**
		 * 1 回分の散らし（TS の scatterOnce）。重なりが出たらその場でリトライし、詰まったら範囲を広げる。
		 * KeepById にある配置はそのまま採用し、そのボクセルは先に占有として登録する。
		 *
		 * 乱数の消費の順が結果を決めるので（RULES.md 3.6）、TS の手順から 1 つも増やさず減らさない。
		 */
		TArray<FPlacement> ScatterOnce(TArrayView<const FPiece> Pieces, int32 N, FRng& Rng, bool bAllowRotation,
			const TMap<int32, FPlacement>& KeepById)
		{
			// 「含むかどうか」だけを見る集合なので TSet でよい（TS は Set<string>）
			TSet<FVec3> Occupied;
			TArray<FPlacement> Placements;
			Placements.Reserve(Pieces.Num());
			// 立方体は [0, N-1]³ にあるので、その中心のまわり ±(N+2) を既定の散らし範囲にする。
			// N >= 3 なので TS の Math.floor((n - 1) / 2) は整数除算でよい
			const int32 Center = (N - 1) / 2;

			// 固定ピースのボクセルを先に占有として登録し、残りがそこへ重ならないようにする
			for (const FPiece& Piece : Pieces)
			{
				const FPlacement* Kept = KeepById.Find(Piece.Id);
				if (Kept == nullptr)
				{
					continue;
				}
				for (const FVec3& V : PlacedVoxels(Piece, *Kept))
				{
					Occupied.Add(V);
				}
			}

			for (const FPiece& Piece : Pieces)
			{
				const FPlacement* Kept = KeepById.Find(Piece.Id);
				if (Kept != nullptr)
				{
					// そのまま残す。乱数も消費しないので keep の内容が同じなら結果も同じ
					Placements.Add(*Kept);
					continue;
				}

				int32 Half = N + 2;
				int32 Attempts = 0;
				for (;;)
				{
					const int32 Span = Half * 2 + 1;
					FPlacement Placement;
					Placement.PieceId = Piece.Id;
					// bAllowRotation が false のときは向きの乱数を引かない（引くと以降の列が全部ずれる）
					Placement.Orientation = bAllowRotation ? Rng.NextInt(OrientationCount) : IdentityOrientation;
					// 引く順は X → Y → Z。C++ は関数の引数の評価順が決まらないので、Vec3(...) の中に直接書かず先に受ける
					const int32 X = Center - Half + Rng.NextInt(Span);
					const int32 Y = Center - Half + Rng.NextInt(Span);
					const int32 Z = Center - Half + Rng.NextInt(Span);
					Placement.Position = Vec3(X, Y, Z);

					const TArray<FVec3> Voxels = PlacedVoxels(Piece, Placement);
					bool bFree = true;
					for (const FVec3& V : Voxels)
					{
						if (Occupied.Contains(V))
						{
							bFree = false;
							break;
						}
					}
					if (bFree)
					{
						for (const FVec3& V : Voxels)
						{
							Occupied.Add(V);
						}
						Placements.Add(Placement);
						break;
					}

					++Attempts;
					checkf(Attempts < ScatterAttemptLimit, TEXT("散らしに失敗: ピース %d を置けなかった"), Piece.Id);
					// 詰まってきたら範囲を広げ、必ず終わるようにする
					if (Attempts % ScatterWidenInterval == 0)
					{
						++Half;
					}
				}
			}
			return Placements;
		}
	}

	int32 MaxPieces(int32 N)
	{
		AssertSpaceSize(N);
		return 5 * N - 8;
	}

	FGeneratedPuzzle GeneratePuzzle(int32 N, int32 M, uint32 Seed)
	{
		AssertSpaceSize(N);
		AssertPieceCount(N, M);
		// TS の assertSeed（シードが整数か）に当たる検査は要らない。シードを uint32 で受けるので型で排除される

		FRng Rng = CreateRng(Seed);
		const int32 Total = N * N * N;
		// Owner[Index] = そのボクセルを持つピース id（未割り当ては Unassigned）
		TArray<int32> Owner;
		Owner.Init(Unassigned, Total);
		// Members[Piece] = そのピースのボクセルのインデックス
		TArray<TArray<int32>> Members;
		Members.SetNum(M);
		// Frontiers[Piece] = 吸収候補（隣接する未割り当てボクセル）。割り当て済みの古い候補も混ざる
		TArray<TArray<int32>> Frontiers;
		Frontiers.SetNum(M);

		/** ボクセルをピースに吸収し、その 6 近傍の空きを候補に積む（TS の absorb） */
		const auto Absorb = [&Owner, &Members, &Frontiers, N](int32 Index, int32 Piece) -> void
		{
			Owner[Index] = Piece;
			Members[Piece].Add(Index);
			const FVec3 V = DecodeIndex(Index, N);
			TArray<int32>& Frontier = Frontiers[Piece];
			for (const FVec3& Offset : NeighborOffsets)
			{
				const FVec3 Neighbor = AddVec3(V, Offset);
				if (!IsInside(Neighbor, N))
				{
					continue;
				}
				const int32 NeighborIndex = EncodeIndex(Neighbor, N);
				if (Owner[NeighborIndex] != Unassigned)
				{
					continue;
				}
				Frontier.Add(NeighborIndex);
			}
		};

		// 2. 相異なるランダム座標に M 個のシードを置く。
		//    部分 Fisher-Yates で先頭 M 個を選ぶので、重複のリトライ無しに必ず相異なる
		TArray<int32> Shuffled;
		Shuffled.Reserve(Total);
		for (int32 Index = 0; Index < Total; ++Index)
		{
			Shuffled.Add(Index);
		}
		for (int32 I = 0; I < M; ++I)
		{
			const int32 J = I + Rng.NextInt(Total - I);
			const int32 A = Shuffled[I];
			const int32 B = Shuffled[J];
			Shuffled[I] = B;
			Shuffled[J] = A;
			Absorb(B, I);
		}

		// 3. 各ピースが隣接する空きを 1 つ吸収する、をラウンドロビンで繰り返す
		int32 Assigned = M;
		while (Assigned < Total)
		{
			bool bGrew = false;
			for (int32 Piece = 0; Piece < M && Assigned < Total; ++Piece)
			{
				TArray<int32>& Frontier = Frontiers[Piece];
				int32 Picked = Unassigned;
				while (Frontier.Num() > 0)
				{
					const int32 K = Rng.NextInt(Frontier.Num());
					const int32 Candidate = Frontier[K];
					// TS と同じ swap-remove（末尾を K に移して末尾を落とす）。
					// この取り出し方が以降の乱数の当たり先を決めるので変えない
					Frontier[K] = Frontier.Last();
					Frontier.RemoveAt(Frontier.Num() - 1);
					if (Owner[Candidate] == Unassigned)
					{
						Picked = Candidate;
						break;
					}
				}
				// 空きが無いピースはこのラウンドをスキップ
				if (Picked == Unassigned)
				{
					continue;
				}
				Absorb(Picked, Piece);
				++Assigned;
				bGrew = true;
			}
			if (bGrew)
			{
				continue;
			}

			// すべてのピースがスキップになったのに空きが残る場合（RULES.md 3.2-3 の孤立した空き）。
			// 候補は吸収されるまで Frontier に残る作りなので実際にはここへ来ないが、仕様どおり保険を置く
			TArray<int32> Remaining;
			for (int32 Index = 0; Index < Total; ++Index)
			{
				if (Owner[Index] == Unassigned)
				{
					Remaining.Add(Index);
				}
			}
			while (Remaining.Num() > 0)
			{
				TArray<int32> Next;
				bool bProgressed = false;
				for (const int32 Index : Remaining)
				{
					const FVec3 V = DecodeIndex(Index, N);
					int32 Host = Unassigned;
					for (const FVec3& Offset : NeighborOffsets)
					{
						const FVec3 Neighbor = AddVec3(V, Offset);
						if (!IsInside(Neighbor, N))
						{
							continue;
						}
						const int32 NeighborOwner = Owner[EncodeIndex(Neighbor, N)];
						if (NeighborOwner != Unassigned)
						{
							Host = NeighborOwner;
							break;
						}
					}
					// どのピースにも隣接しない空きは次の周回に回す（周りが埋まれば吸収できる）
					if (Host == Unassigned)
					{
						Next.Add(Index);
						continue;
					}
					Absorb(Index, Host);
					++Assigned;
					bProgressed = true;
				}
				checkf(bProgressed, TEXT("生成に失敗: どのピースにも隣接しない空きが残った"));
				Remaining = MoveTemp(Next);
			}
		}

		// 4. 解答位置（絶対座標）から局所座標へ正規化し、正規化で引いたオフセットを解答配置にする
		FGeneratedPuzzle Puzzle;
		Puzzle.N = N;
		Puzzle.M = M;
		Puzzle.Seed = Seed;
		Puzzle.Pieces.Reserve(M);
		Puzzle.Solution.Reserve(M);
		for (int32 Id = 0; Id < M; ++Id)
		{
			TArray<FVec3> Absolute;
			Absolute.Reserve(Members[Id].Num());
			for (const int32 Index : Members[Id])
			{
				Absolute.Add(DecodeIndex(Index, N));
			}
			// JS の Array.prototype.sort は安定ソート（Docs/SPEC_UE.md 7.1）。
			// CompareVec3 は相異なる座標の全順序なので、結果の並びは一意に決まる
			Algo::StableSort(Absolute, [](const FVec3& A, const FVec3& B) -> bool
			{
				return CompareVec3(A, B) < 0;
			});

			Puzzle.Pieces.Add(CreatePiece(Id, Absolute));

			FPlacement Placement;
			Placement.PieceId = Id;
			Placement.Orientation = IdentityOrientation;
			Placement.Position = LocalOrigin(Absolute);
			Puzzle.Solution.Add(Placement);
		}

		return Puzzle;
	}

	TArray<FPiece> GeneratePieces(int32 N, int32 M, uint32 Seed)
	{
		// 呼び出しごとに作り直すので、返した配列を書き換えても次の呼び出しには影響しない（TS の .slice()）
		FGeneratedPuzzle Puzzle = GeneratePuzzle(N, M, Seed);
		return MoveTemp(Puzzle.Pieces);
	}

	TArray<FPlacement> SolutionPlacements(int32 N, int32 M, uint32 Seed)
	{
		FGeneratedPuzzle Puzzle = GeneratePuzzle(N, M, Seed);
		return MoveTemp(Puzzle.Solution);
	}

	TArray<FPlacement> ScatterPlacements(TArrayView<const FPiece> Pieces, int32 N, uint32 Seed, const FScatterOptions& Options)
	{
		AssertSpaceSize(N);
		// TS の assertSeed（シードが整数か）に当たる検査は要らない。シードを uint32 で受けるので型で排除される
		checkf(Pieces.Num() > 0, TEXT("散らし: ピースが空"));

		const TMap<int32, FPlacement> KeepById = BuildKeepMap(Pieces, Options.Keep);

		// 呼び出しごとに作り直すので、散らしと散らし直しは互いに乱数を食い合わせない（Docs/FIXTURES.md）
		FRng Rng = CreateRng(Seed);
		TArray<FPlacement> Last;
		bool bHasLast = false;
		// 引き直しは同じ Rng を使い続ける（乱数の続きを消費する）
		for (int32 Retry = 0; Retry < ScatterRetryLimit; ++Retry)
		{
			TArray<FPlacement> Placements = ScatterOnce(Pieces, N, Rng, Options.bAllowRotation, KeepById);
			if (!IsSolved(Pieces, Placements, N))
			{
				return Placements;
			}
			Last = MoveTemp(Placements);
			bHasLast = true;
		}

		checkf(KeepById.Num() > 0 && bHasLast, TEXT("散らしに失敗: クリア状態でない配置を作れなかった"));
		return Last;
	}
}
